### ISSUE TITLE
static mace lib can not be used

### ISSUE THREAD

**Issue by eudora-jia** [XiaoMi/mace#534]
Before you open an issue, please make sure you have tried the following steps:

1. Make sure your **linux** is the same with (https://mace.readthedocs.io/en/latest/installation/env_requirement.html).
2. Have you ever read the document for your usage?
3. Check if your issue appears in [HOW-TO-DEBUG](https://mace.readthedocs.io/en/latest/development/how_to_debug.html) or [FAQ](https://mace.readthedocs.io/en/latest/faq.html).
4. The form below must be filled.

------------------------

### System information
- **Linux localhost.localdomain 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux**:
- **16b**:
- **cmake**:
- **mace master**:
- **Python version(2.7)**: 
- **Bazel version (e.g., 0.13.0)**:

### Model deploy file (*.yml)
```yaml
......
```

### Describe the problem
When I download the repository,I built the source by bazel as the mace.readthedocs.io says, after that,I got two libs which included libmace.so and libmace.a; I used the dynamic lib in my mylib.so with no errors, I ran my app with mylib.so successfully,for safety try,I need to use static mylib.a, So I rebuilt mylib as static one,I met  errors: many many undefined reference;
### To Reproduce
just move the libmace.a to your compiler env, and build static mylib.a,  after  building ,use the mylib to build app, you can get the errors in android studio;
```

## app built with mylib.a which built with libmace.a,Error information / logs
##app built log like this:

/proc/self/cwd/mace/ops/registry/ops_registry.cc:95: undefined reference to `mace::ops::RegisterActivation(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:96: undefined reference to `mace::ops::RegisterAddN(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:98: undefined reference to `mace::ops::RegisterBatchNorm(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:99: undefined reference to `mace::ops::RegisterBatchToSpaceND(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:100: undefined reference to `mace::ops::RegisterBiasAdd(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:102: undefined reference to `mace::ops::RegisterChannelShuffle(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:103: undefined reference to `mace::ops::RegisterConcat(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:104: undefined reference to `mace::ops::RegisterConv2D(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:105: undefined reference to `mace::ops::RegisterCrop(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:107: undefined reference to `mace::ops::RegisterDeconv2D(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:108: undefined reference to `mace::ops::RegisterDepthToSpace(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:109: undefined reference to `mace::ops::RegisterDepthwiseConv2d(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:110: undefined reference to `mace::ops::RegisterDepthwiseDeconv2d(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:112: undefined reference to `mace::ops::RegisterEltwise(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:116: undefined reference to `mace::ops::RegisterFullyConnected(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:124: undefined reference to `mace::ops::RegisterMatMul(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:126: undefined reference to `mace::ops::RegisterPad(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:129: undefined reference to `mace::ops::RegisterPooling(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:130: undefined reference to `mace::ops::RegisterReduce(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:133: undefined reference to `mace::ops::RegisterResizeBicubic(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:134: undefined reference to `mace::ops::RegisterResizeBilinear(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:135: undefined reference to `mace::ops::RegisterResizeNearestNeighbor(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:140: undefined reference to `mace::ops::RegisterSoftmax(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:141: undefined reference to `mace::ops::RegisterSpaceToBatchND(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:142: undefined reference to `mace::ops::RegisterSpaceToDepth(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:144: undefined reference to `mace::ops::RegisterSplit(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:147: undefined reference to `mace::ops::RegisterSqrDiffMean(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:156: undefined reference to `mace::ops::RegisterDequantize(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:157: undefined reference to `mace::ops::RegisterQuantize(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:161: undefined reference to `mace::ops::RegisterBufferTransform(mace::OpRegistryBase*)'
/proc/self/cwd/mace/ops/registry/ops_registry.cc:162: undefined reference to `mace::ops::RegisterLSTMCell(mace::OpRegistryBase*)'

### Additional context
finally I checked your cmakefiles.txt, It turned out that your cmake is just for something simple:
I guess the doings can not be used to prove it walk anywhere!
`file(GLOB MACE_RUN_SRCS
  mace_run.cc
)
add_executable(mace_run ${MACE_RUN_SRCS})
target_link_libraries(mace_run PUBLIC
  mace
  gflags
)`

**Comment by lu229** [comment]
@eudora-jia Please include the full log, I will check the problem

**Comment by eudora-jia** [comment]
I add that now;actual I do like this:
**#这个脚本才可以真正的生成静态库
set -e

LIB=libfingertip
ANDROID_ABI=arm64-v8a
READ_OPT=LibReadFile
AR=${HOME}/tools/android-ndk-r16b/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/bin/aarch64-linux-android-ar
RANLIB=${HOME}/tools/android-ndk-r16b/toolchains/aarch64-linux-android-4.9/prebuilt/linux-x86_64/bin/aarch64-linux-android-ranlib

rm -rf /tmp/$LIB
mkdir /tmp/$LIB
cd /tmp/$LIB

$AR x $HOME/work/fingertip/3rdparty/mace-master/lib/$ANDROID_ABI/cpu_gpu/libmace.a
#$AR x $HOME/work/fingertip/install/lib/$ANDROID_ABI/$READ_OPT/"$LIB"_static.a 
#$AR vqc $LIB.a *.o

#$RANLIB -v $LIB.a
**

**Comment by eudora-jia** [comment]
actually, I strings my built static mylib.a, I got the symbol mace::ops::RegisterActivation(mace::OpRegistryBase*) ,  so why, and I check the sort of linking, I got this:
`[myname@172-10-10-157 fingertip]$ ./build_static.sh 
a - activation.o
a - addn.o
a - airia_engine.cpp.o
a - allocator.o
a - arena.o
a - arenastring.o
a - arg_helper.o
a - argmax.o
a - batch_norm.o
a - batch_to_space.o
a - bias_add.o
a - buffer_to_image.o
a - buffer_transformer.o
a - buffer_transform.o
a - buffer_type_transform.o
a - bytestream.o
a - capability.o
a - cast.o
a - channel_shuffle.o
a - coded_stream.o
a - common.cpp.o
a - common.o
a - concat.o
a - conv_2d_1x1.o
a - conv_2d_1xn.o
a - conv_2d_3x3.o
a - conv_2d_3x3_winograd.o
a - conv_2d_5x5.o
a - conv_2d_7x7.o
a - conv_2d_general.o
a - conv_2d.o
a - conv_general.o
a - conv_pool_2d_util.o
a - cpu_runtime.o
a - crop.o
a - cumsum.o
a - deconv_2d_2x2.o
a - deconv_2d_3x3.o
a - deconv_2d_4x4.o
a - deconv_2d_general.o
a - deconv_2d.o
a - delay.o
a - depth_to_space.o
a - depthwise_conv_2d_3x3.o
a - depthwise_conv_2d.o
a - depthwise_conv2d.o
a - depthwise_deconv_2d_3x3.o
a - depthwise_deconv_2d_4x4.o
a - depthwise_deconv_2d_general.o
a - depthwise_deconv_2d.o
a - depthwise_deconv2d.o
a - device_context.o
a - device.o
a - dynamic_lstm.o
a - eltwise.o
a - encrypt_opencl_kernel.o
a - env.o
a - expand_dims.o
a - extension_set.o
a - extract_pooling.o
a - file_system.o
a - fill.o
a - fingertips_det.cpp.o
a - fully_connected.o
a - gather.o
a - gemm.o
a - gemv.o
a - generated_message_table_driven_lite.o
a - generated_message_util.o
a - gpu_device.o
a - gpu_runtime.o
a - helper.o
a - identity.o
a - image_to_buffer.o
a - implicit_weak_message.o
a - infer_conv2d_shape.o
a - int128.o
a - io_win32.o
a - kaldi_batch_norm.o
a - kv_storage.o
a - local_response_norm.o
a - logger.o
a - lstm_cell.o
a - lstm_nonlinear.o
a - lstm.o
a - mace.o
a - mace.pb.o
a - malloc_logger.o
a - matmul.o
a - memory_optimizer.o
a - message_lite.o
a - net_def_adapter.o
a - net.o
a - net_optimizer.o
a - one_hot.o
a - op_context.o
a - opencl_allocator.o
a - opencl_runtime.o
a - opencl_util.o
a - opencl_wrapper.o
a - operator.o
a - ops_registry.o
a - pad_context.o
a - pad.o
a - pnorm.o
a - pooling.o
a - prior_box.o
a - quantization_util.o
a - quantize.o
a - reduce.o
a - repeated_field.o
a - reshape.o
a - resize_bicubic.o
a - resize_bilinear.o
a - resize_nearest_neighbor.o
a - reverse.o
a - runtime_failure_mock.o
a - scalar_math.o
a - scratch_image.o
a - shape.o
a - slice.o
a - softmax.o
a - space_to_batch.o
a - space_to_depth.o
a - splice.o
a - split.o
a - sqrdiff_mean.o
a - squeeze.o
a - stack.o
a - statistics.o
a - status.o
a - statusor.o
a - strided_slice.o
a - stringpiece.o
a - stringprintf.o
a - string_util.o
a - structurally_valid.o
a - strutil.o
a - sum_group.o
a - target_rms_norm.o
a - thread_pool.o
a - time.o
a - transpose.o
a - types.o
a - unsqueeze.o
a - unstack.o
a - utils.o
a - version.o
a - winograd_conv2d.o
a - wire_format_lite.o
a - workspace.o
a - zero_copy_stream_impl_lite.o
a - zero_copy_stream.o
`

**Comment by lu229** [comment]
@eudora-jia Please fill in the [issue template](https://github.com/XiaoMi/mace/blob/master/ISSUE_TEMPLATE.md), otherwise we don't know what's your problem.
